### PR TITLE
fix: improve autofix of `prefer-object-has-own`

### PR DESCRIPTION
**Pull request by snitin315**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Resolve https://github.com/eslint/eslint/pull/15346#issuecomment-991417335

Improve autofix for the case - `return{}.hasOwnProperty.call(object, property)`

#### Is there anything you'd like reviewers to focus on?
No

### CODE REVIEW

File: lib/rules/prefer-object-has-own.js [hunk 1]
```diff
@@ -87,6 +87,13 @@ module.exports = {
                                 return null;
                             }
 
+                            const tokenJustBeforeNode = sourceCode.getTokensBefore(node, 1)[0];
```
Comment (fisker) [inline lib/rules/prefer-object-has-own.js:90]: There is a `getTokenBefore` method.

We can also use `includeComments`. if there is comment between, the space it not nessary.

File: lib/rules/prefer-object-has-own.js [hunk 2]
```diff
@@ -87,6 +87,13 @@ module.exports = {
...
+
+                            // for https://github.com/eslint/eslint/pull/15346#issuecomment-991417335
+                            if (tokenJustBeforeNode && tokenJustBeforeNode.type === "Keyword" && !sourceCode.isSpaceBetween(node, tokenJustBeforeNode)) {
```
Comment (fisker) [inline lib/rules/prefer-object-has-own.js:93]: This is not enough, see https://github.com/sindresorhus/eslint-plugin-unicorn/blob/32c20196528c28d9dd6eb1f3beb15f5528480418/rules/fix/fix-space-around-keywords.js#L4

Comment (snitin315) [reply lib/rules/prefer-object-has-own.js]: Thanks.

File: lib/rules/prefer-object-has-own.js [hunk 3]
```diff
@@ -36,6 +40,26 @@ function hasLeftHandObject(node) {
...
+ * @returns {boolean} `true` if the token is  is `await` Identefier, `of` Itentifier or a keyword token
+ */
+function isProblematicToken(token) {
+    const { type, value } = token;
+
+    if (
+        (type === "Keyword" && /^[a-z]*$/u.test(value)) ||
+
+        // AwaitExpression or ForOfStatement
+        (type === "Identifier" && (value === "of" || value === "await"))) {
```
Comment (fisker) [inline lib/rules/prefer-object-has-own.js:56]: `of` can also be tested

```js
for (const x of{}.hasOwnProperty.call(object, property));
```

If you want it make more sense, convert boolean to string

```js
for (const x of{}.hasOwnProperty.call(object, property).toString());
```

Comment (snitin315) [reply lib/rules/prefer-object-has-own.js:56]: Thanks, I added more test cases.

**Review by fisker — APPROVED** [review]

File: lib/rules/prefer-object-has-own.js [hunk 4]
```diff
@@ -87,6 +111,13 @@ module.exports = {
                                 return null;
                             }
 
+                            const tokenJustBeforeNode = sourceCode.getTokenBefore(node, { includeComments: true });
```
Comment (mdjermanovic) [inline lib/rules/prefer-object-has-own.js:114]: This example:

```js
function foo() {
    return({}.hasOwnProperty.call)(obj, prop);
}
```

would be autofixed with an unnecessary space:

```js
function foo() {
    return( Object.hasOwn)(obj, prop);
}
```

We should get token before the node that is going to be replaced, which is `node.callee`.

```suggestion
                            const tokenJustBeforeNode = sourceCode.getTokenBefore(node.callee, { includeComments: true });
```

File: lib/rules/prefer-object-has-own.js [hunk 5]
```diff
@@ -87,6 +111,13 @@ module.exports = {
...
+
+                            // for https://github.com/eslint/eslint/pull/15346#issuecomment-991417335
+                            if (tokenJustBeforeNode && isProblematicToken(tokenJustBeforeNode) && !sourceCode.isSpaceBetween(node, tokenJustBeforeNode)) {
```
Comment (mdjermanovic) [inline lib/rules/prefer-object-has-own.js:117]: We're using [`astUtils.canTokensBeAdjacent()`](https://github.com/eslint/eslint/blob/eafaf52464c01072fcf7be2bd2d91cd0b3ae67d0/lib/rules/utils/ast-utils.js#L1744) for this purpose.

An example:

https://github.com/eslint/eslint/blob/eafaf52464c01072fcf7be2bd2d91cd0b3ae67d0/lib/rules/no-implicit-coercion.js#L241-L250

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: lib/rules/prefer-object-has-own.js (hunk 5 above)
Comment (snitin315) [reply lib/rules/prefer-object-has-own.js]: Updated to code to use `astUtils.canTokensBeAdjacent() ` 👍🏻

**Review by mdjermanovic — COMMENTED** [review]
This works well, I have only one small suggestion.

File: lib/rules/prefer-object-has-own.js [hunk 6]
```diff
@@ -87,6 +91,17 @@ module.exports = {
                                 return null;
                             }
 
+                            const tokenJustBeforeNode = sourceCode.getTokenBefore(node.callee, { includeComments: true });
+
+                            // for https://github.com/eslint/eslint/pull/15346#issuecomment-991417335
+                            if (
+                                tokenJustBeforeNode &&
+                                !astUtils.canTokensBeAdjacent(tokenJustBeforeNode, "Object.hasOwn") &&
+                                !sourceCode.isSpaceBetween(node.callee, tokenJustBeforeNode)
```
[the remark this answers is not in the document]
Comment (snitin315) [reply lib/rules/prefer-object-has-own.js]: yes, updated 👍🏻

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**